lib/CMakeLists.txt

```diff
@@ -153,183 +154,74 @@ endif()
 add_library(quda_cpp OBJECT ${QUDA_OBJS})
 
 # add comms and QIO
-target_sources(quda_cpp PRIVATE $<IF:$<BOOL:${QUDA_MPI}>,communicator_mpi.cpp,$<IF:$<BOOL:${QUDA_QMP}>,communicator_qmp.cpp,communicator_single.cpp>>)
+target_sources(
+  quda_cpp
+  PRIVATE
+    $<IF:$<BOOL:${QUDA_MPI}>,communicator_mpi.cpp,$<IF:$<BOOL:${QUDA_QMP}>,communicator_qmp.cpp,communicator_single.cpp>>
+)
 
 target_sources(quda_cpp PRIVATE $<$<BOOL:${QUDA_QIO}>:qio_field.cpp layout_hyper.cpp>)
 
-# add some deifnitions that cause issues with cmake 3.7 and nvcc only to cpp files
-target_compile_definitions(quda_cpp PUBLIC -DQUDA_HASH="${HASH}")
-if(GITVERSION)
-  target_compile_definitions(quda_cpp PUBLIC -DGITVERSION="${GITVERSION}")
-endif()
-
-# make one library
 if(QUDA_BUILD_SHAREDLIB)
   set_target_properties(quda_cpp PROPERTIES POSITION_INDEPENDENT_CODE TRUE)
-  add_library(quda SHARED $<TARGET_OBJECTS:quda_cpp> $<$<TARGET_EXISTS:quda_pack>:$<TARGET_OBJECTS:quda_pack>> ${QUDA_CU_OBJS})
-  if(CUDAToolkit_FOUND)
-    target_link_libraries(quda INTERFACE CUDA::cudart_static)
-  endif()
+  add_library(quda SHARED)
 else()
-  add_library(quda STATIC $<TARGET_OBJECTS:quda_cpp> $<$<TARGET_EXISTS:quda_pack>:$<TARGET_OBJECTS:quda_pack>> ${QUDA_CU_OBJS})
+  add_library(quda STATIC)
+endif()
+add_library(QUDA::quda ALIAS quda)
+
+# add target specific files
+if(${QUDA_TARGET_TYPE} STREQUAL "CUDA")
+  include(targets/cuda/target_cuda.cmake)
+endif()
+if(${QUDA_TARGET_TYPE} STREQUAL "HIP")
+  add_subdirectory(targets/hip)
+  target_include_directories(quda PRIVATE ../include/targets/hip)
+  set(QUDA_TARGET_HIP ON)
 endif()
 
```

Comment:
I will need to make the HIP target similar to the CUDA target here, but that will be my problem...